common/src/main/java/com/viaversion/viaversion/rewriter/RegistryDataRewriter.java

```diff
@@ -65,11 +67,23 @@ public RegistryEntry[] handle(final UserConnection connection, String key, Regis
 
         final List<RegistryEntry> toAdd = this.toAdd.get(key);
         if (toAdd != null) {
-            final int length = entries.length;
-            final int toAddLength = toAdd.size();
-            entries = Arrays.copyOf(entries, length + toAddLength);
-            for (int i = 0; i < toAddLength; i++) {
-                entries[length + i] = toAdd.get(i).copy();
+            final Set<String> toAddKeys = toAdd.stream().map(RegistryEntry::key).collect(Collectors.toSet());
+
+            final RegistryEntry[] updatedEntries = new RegistryEntry[entries.length + toAdd.size()];
+            int index = 0;
+            for (final RegistryEntry entry : entries) {
+                if (!toAddKeys.contains(entry.key())) {
+                    updatedEntries[index++] = entry.copy();
```

Comment:
The original entries don't need copying, but the server-sent entry should also be preferred over what we add

Reply:
Don't we only add registry elements that older clients don't have and thus older servers don't send? In these cases it's wrong to prefer the server-sent as a Vanilla client in that version would ignore them.